// figgy-pudding/index.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var FiggyPudding = function () {
  function FiggyPudding(specs, opts, providers) {
    _classCallCheck(this, FiggyPudding);

    this.specs = specs || {};
    this.opts = opts || function () {
      return false;
    };
    this.providers = providers;
    this.isFiggyPudding = true;
  }

  _createClass(FiggyPudding, [{
    key: 'get',
    value: function get(key) {
      return pudGet(this, key, true);
    }
  }]);

  return FiggyPudding;
}();

function pudGet(pud, key, validate) {
  var spec = pud.specs[key];
  if (typeof spec === 'string') {
    key = spec;
    spec = pud.specs[key];
  }
  if (validate && !spec && (!pud.opts.other || !pud.opts.other(key))) {
    throw new Error('invalid config key requested: ' + key);
  } else {
    if (!spec) {
      spec = {};
    }
    var ret = void 0;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = pud.providers[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var p = _step.value;

        if (p.isFiggyPudding) {
          ret = pudGet(p, key, false);
        } else if (typeof p.get === 'function') {
          ret = p.get(key);
        } else {
          ret = p[key];
        }
        if (ret !== undefined) {
          break;
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    if (ret === undefined && spec.default !== undefined) {
      if (typeof spec.default === 'function') {
        return spec.default();
      } else {
        return spec.default;
      }
    } else {
      return ret;
    }
  }
}

module.exports = figgyPudding;
function figgyPudding(specs, opts) {
  function factory() {
    return new FiggyPudding(specs, opts, [].slice.call(arguments).filter(function (x) {
      return x != null && (typeof x === 'undefined' ? 'undefined' : _typeof(x)) === 'object';
    }));
  }
  return factory;
}
