// libnpmhook/index.js
'use strict';

var config = require('./config');
var fetch = require('npm-registry-fetch');

module.exports = {
  add: function add(name, endpoint, secret, opts) {
    var type = 'package';
    if (name && name.match(/^@[^/]+$/)) {
      type = 'scope';
    }
    if (name && name[0] === '~') {
      type = 'owner';
      name = name.substr(1);
    }

    opts = config({
      method: 'POST',
      body: { type: type, name: name, endpoint: endpoint, secret: secret }
    }, opts);
    return fetch.json('/-/npm/v1/hooks/hook', opts);
  },
  rm: function rm(id, opts) {
    return fetch.json('/-/npm/v1/hooks/hook/' + encodeURIComponent(id), config({
      method: 'DELETE'
    }, opts));
  },
  ls: function ls(pkg, opts) {
    return fetch.json('/-/npm/v1/hooks', config({ query: pkg && { package: pkg } }, opts)).then(function (json) {
      return json.objects;
    });
  },
  update: function update(id, endpoint, secret, opts) {
    return fetch.json('/-/npm/v1/hooks/hook/' + encodeURIComponent(id), config({
      method: 'PUT',
      body: { endpoint: endpoint, secret: secret }
    }, opts));
  }
};
